src/control/server/instance_drpc.go

```diff
@@ -215,86 +216,89 @@ func (ei *EngineInstance) getSmdDetails(smd *ctlpb.SmdDevResp_Device) (*storage.
 // SMD metadata info.
 //
 // Query each SmdDevice on each I/O Engine instance for health stats and update existing controller
 // data in ctrlrMap using PCI address key.
-func (ei *EngineInstance) updateInUseBdevs(ctx context.Context, ctrlrMap map[string]*storage.NvmeController) error {
+func (ei *EngineInstance) updateInUseBdevs(ctx context.Context, ctrlrMap map[string]*storage.NvmeController) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "instance %d", ei.Index())
+	}()
+
+	// Clear smd info for controllers in ctrlrMap, populate smd info from scratch.
+	for _, ctrlr := range ctrlrMap {
+		ctrlr.SmdDevices = []*storage.SmdDevice{}
+	}
+
 	smdDevs, err := ei.ListSmdDevices(ctx, new(ctlpb.SmdDevReq))
 	if err != nil {
-		return errors.Wrapf(err, "instance %d listSmdDevices()", ei.Index())
+		return errors.Wrapf(err, "list smd devices")
 	}
+	ei.log.Debugf("engine %d: smdDevs %+v", ei.Index(), smdDevs)
 
 	hasUpdatedHealth := make(map[string]bool)
 	for _, smd := range smdDevs.Devices {
+		msg := fmt.Sprintf("instance %d: smd %s: ctrlr %s", ei.Index(), smd.Uuid,
+			smd.TrAddr)
+
 		ctrlr, exists := ctrlrMap[smd.GetTrAddr()]
 		if !exists {
-			return errors.Errorf("instance %d: smd %s: unknown controller %s",
-				ei.Index(), smd.GetUuid(), smd.GetTrAddr())
+			ei.log.Errorf("%s: ctrlr not found", msg)
+			continue
 		}
 
 		smdDev, err := ei.getSmdDetails(smd)
 		if err != nil {
-			return errors.Wrapf(err, "collect smd info for ctrlr %s", ctrlr.PciAddr)
+			ei.log.Errorf("%s: collect smd info: %s", msg, err.Error())
+			continue
+		}
+
+		pbStats, err := ei.GetBioHealth(ctx, &ctlpb.BioHealthReq{DevUuid: smdDev.UUID})
+		if err != nil {
+			// continue if error indicates non-existent health and smd has abnormal state
+			status, ok := errors.Cause(err).(drpc.DaosStatus)
+			if ok && status == drpc.DaosNonexistant && !smdDev.NvmeState.IsNormal() {
+				ei.log.Debugf("%s: stats not found (device state: %q), skip update",
+					msg, smdDev.NvmeState.String())
+			} else {
+				ei.log.Errorf("%s: fetch stats: %s", msg, err.Error())
+			}
+			ctrlr.UpdateSmd(smdDev)
+			continue
 		}
 
+		// populate space usage for each smd device from health stats
+		smdDev.TotalBytes = pbStats.TotalBytes
+		smdDev.AvailBytes = pbStats.AvailBytes
+		msg = fmt.Sprintf("%s: smd usage = %s/%s", msg, humanize.Bytes(smdDev.AvailBytes),
+			humanize.Bytes(smdDev.TotalBytes))
+		ctrlr.UpdateSmd(smdDev)
+
 		// multiple updates for the same key expected when more than one controller
 		// namespaces (and resident blobstores) exist, stats will be the same for each
 		// so only pass valid ctrlr reference when stats haven't yet been updated
-		var ctrlrRef *storage.NvmeController
-		if _, already := hasUpdatedHealth[ctrlr.PciAddr]; !already {
-			ctrlrRef = ctrlr
+		if hasUpdatedHealth[ctrlr.PciAddr] {
+			ei.log.Debugf("%s: health stats already added so skip update", msg)
```

Comment:
Likewise, this seems unnecessarily chatty for landing.

Reply:
as above, this is unlikely to be very frequent

Reply:
I'd think if a machine has one drive with multiple namespaces, though, it'd be likely they'd all have multiple namespaces. Is that a bad assumption? On a machine with a lot of drives, I could see this one getting noisy.

Reply:
from what I've seen it's rare to have multiple namespaces on a NVMe controller, common to just have a single ns

Reply:
removed